### ISSUE TITLE
【OPD】Support student-top-k renormalized reverse-KL OPD

### ISSUE THREAD

**Issue by zsychina** [verl-project/verl#6676]
### Feature request

Could verl support the student-top-k Top-k OPD formulation as an additional loss mode?

From the current implementation, `forward_kl_topk` appears to use teacher top-k support and compute a forward-KL-like objective on an unnormalized subset. This differs from the student-top-k OPD formulation:
$$S_t = TopK(p_t, k)$$, followed by $$KL(p_t^{(S_t)} || q_t^{(S_t)})$$.

I would like to request either:
1. a new loss mode for student-top-k renormalized reverse-KL OPD, or
2. documentation clarifying that current `forward_kl_topk` is not equivalent to that formulation.

Top-k OPD is from paper "Rethinking On-Policy Distillation of Large Language Models: Phenomenology, Mechanism, and Recipe" https://arxiv.org/abs/2604.13016

<img width="820" height="397" alt="Image" src="https://github.com/user-attachments/assets/32b39eea-58d0-4433-bfdc-77bd74ba2c1d" />

### Motivation

to support papper rethinking opd

### Your contribution

no pr planned

**Comment by Fyrgo8** [comment]
Hi @zsychina, I'd like to pick this up. I've read through the current OPD implementation and want to align on the formulation before writing any code.

**Confirming the gap.** I'll use π for the student and ν for the teacher to keep directions unambiguous. The current `forward_kl_topk` gathers student log-probs on the teacher's top-k ids and computes Σ_{i∈T} ν_i (log ν_i − log π_i), i.e. an unnormalized **forward** KL over the **teacher** top-k. What this issue asks for is the opposite on both axes — a **reverse** KL renormalized over the **student** top-k S = TopK(π, k). So they genuinely differ in both KL direction and top-k owner; I'll make sure option 2 (a docs clarification) is covered regardless, since `docs/algo/opd.md` currently only states the non-equivalence without explaining the renormalization obstacle below.

**The core obstacle to the exact renormalized objective.** The target is

> L* = Σ_{i∈S} π̃_i · log(π̃_i / ν̃_i),  with ν̃_i = ν_i / Σ_{j∈S} ν_j.

Both the numerator ν_i for i ∈ S\T and the renormalization constant Σ_{j∈S} ν_j require teacher probabilities on the **student's** top-k ids. But the teacher is served with `prompt_logprobs=topk`, which only returns logprobs on the teacher's own top-k — it can't gather arbitrary token ids — and teacher inference runs before the student forward, so we can't go back and ask for S. That makes L* exactly uncomputable as-is; any implementation is necessarily an approximation, and I think it's worth being upfront about that in both the code and the docs.

**Two landable approximations I'd propose, with a preference.**

**(A) Unnormalized reverse KL over the student top-k with a tail backoff (my inclination).** Compute Σ_{i∈S} π_i (log π_i − log ν̂_i), using the true ν_i for i ∈ S∩T and a backoff ν̂_i for i ∈ S\T. One subtlety worth flagging: the loss carries −log ν̂_i, so the conservative direction (not under-counting divergence) requires ν̂_i to be a **lower** bound on ν_i. This rules out 1 − m_ν (the total tail mass — an upper bound on any single ν_i), which would systematically under-estimate the loss, the wrong direction for a distillation penalty. It also means the uniform spread ν̂_i = (1 − m_ν)/(|V| − k) is not strictly conservative either: it's a point estimate, and for tokens whose true ν_i falls **below** the tail average it over-estimates ν̂_i and thus **under-counts** the loss (a larger ν̂_i means a smaller −log ν̂_i penalty). So there's a genuine choice to make here: a clamped floor via the existing `log_prob_min_clamp` if we want a stronger conservative guarantee, or the uniform spread if we'd rather have a lower-variance point estimate and accept that it isn't a strict bound. I lean toward the clamped floor as the default — note it only keeps the loss an upper bound when the floor e^c ≤ ν_i, so for tokens whose true ν_i is already below e^c the clamp raises ν̂_i above ν_i and the bound no longer holds; in practice c is small enough that this is rare, but it isn't an unconditional guarantee. Happy to be overruled if you prefer the point estimate. Either way I'd deliberately avoid a pure S∩T intersection mask, because that drops exactly the S\T tokens — the ones where the student puts mass the teacher doesn't, which reverse KL should penalize most.

**(B) Intersection-renormalized reverse KL + backoff on the rest.** Renormalize the available teacher probs over S∩T (ν̃_i = ν_i / Σ_{j∈S∩T} ν_j) and keep an explicit backoff term for S\T rather than masking it out.

**Making the (A)/(B) choice decidable.** The substantive difference between them is whether the teacher probabilities get renormalized over S∩T, and correspondingly whether the student side is aligned onto the same support. In (A) I keep both π_i and ν_i as their raw (full-softmax-derived) values and let the S\T backoff carry the residual mass — so (A) is an unnormalized reverse KL where the only modeling choice is the backoff for the teacher tail. In (B) I divide by Σ_{j∈S∩T} ν_j to form a proper teacher distribution on the intersection, which is closer to the renormalization you asked for but forces a decision on how the student side is normalized over the same support (otherwise the two distributions in the KL no longer share a support and the divergence is ill-posed). Everything else — student top-k selection, the extra gather, the conservative-backoff requirement — is identical between the two. So the question I'm really asking is narrow: do you want teacher-side renormalization on S∩T (B), or an unnormalized objective with a tail backoff (A)?

**Implementation & PR split.** On FSDP this reuses the `student_topk_ids` already computed in the logits processor, plus one extra gather of student log-probs on S (the current code only gathers on the teacher top-k). I'd split into two PRs:

- **PR 1 (FSDP only)** — the new `reverse_kl_topk` loss mode + a CPU unit test that asserts numerical equality to a naive full-vocab reverse KL in the degenerate construction where S ⊆ T (so the backoff never fires), plus the docs clarification.
- **PR 2 (Megatron follow-up)** — the Megatron path is a hand-written vocab-parallel `autograd.Function` whose backward is analytically derived for forward KL (dL/dz_j = m_A · p_j − q_j); reverse KL changes the per-token loss to π_i (log π_i − log ν_i) and needs the backward re-derived and validated multi-GPU, which is higher-risk and better landed separately.

**On validation beyond the degenerate test.** The S ⊆ T unit test only pins down the easy case where the backoff never fires, which isn't the regime that actually matters — in practice S ⊄ T almost always, and that's exactly where the approximation error lives and where I currently have zero data. So in the draft PR I'll include a small toy-scale numerical study (random teacher/student logit pairs, no training) that compares the approximate loss against the exact full-vocab renormalized reverse KL L*, sweeping k and the S∩T overlap ratio, and reports the bias as a function of overlap. That turns "this is a principled approximation" into something falsifiable — we'd see directly how fast the error decays as overlap grows and whether the chosen backoff stays on the conservative side empirically. If you'd prefer, I can post those curves in the issue before the PR so the approximation semantics are settled with numbers, not just the analysis above.

**One question to settle first:** which approximation semantics do you want — (A) unnormalized + tail backoff, or (B) intersection-renormalized + backoff? And is FSDP-first with Megatron as a follow-up acceptable, or would you prefer both backends in one PR?

Happy to open a draft PR early once the formulation is settled. Thanks!

**Comment by Fyrgo8** [comment]
Hi @wuxibin89 — would appreciate your steer here. I've posted a detailed proposal above; the two open questions are (1) approximation semantics: (A) unnormalized reverse KL over the student top-k with a conservative tail backoff, or (B) intersection-renormalized reverse KL; and (2) whether FSDP-first with a Megatron follow-up is acceptable, or you'd prefer both backends in one PR. Leaning toward (A) + FSDP-first. Happy to open a draft once the direction's confirmed.
